pkg/ostree/pre_run_action.go

```diff
@@ -0,0 +1,117 @@
+package ostree
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+	"k8s.io/klog/v2"
+)
+
+type action string
+
+const (
+	actionBackup  action = "backup"
+	actionRestore action = "restore"
+	actionMissing action = "missing"
+
+	persistenceFile = "pre_run_action"
+	filePerm        = os.FileMode(0644)
+)
+
+var preRunActionFilepath = filepath.Join(config.AuxDataDir, persistenceFile)
+
+var getFileWriter = func() (io.Writer, error) {
```

Comment:
I don't have strong preference, even what I implemented already is not elegant in my eyes.
What I aimed to achieve to have untested layer as thin as possible. If we add more explicit dependency injection or ActionManager interface with different impls I feel like we're going away from the filesystem and adding more bug vectors.
About passing io.Reader/Writer - current way is not that far, but does it in a different way. Actually, taking `Persist()` as an example: would it result in just taking code from json.Marshal  to the end to another function that would get io.Writer? Sort of like:
```go
func (nb *preRunAction) Persist() error {
	if err := config.EnsureAuxDirExists(); err != nil {
		return err
	}
	w, err := os.OpenFile(preRunActionFilepath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, filePerm)
	if err != nil {
		return err
	}
	return nb.persist(w)
}

func (nb *preRunAction) persist(w io.Writer) error {
	b, err := json.Marshal(nb)
	if err != nil {
		return err
	}
	n, err := w.Write(b)
	if err != nil {
		return err
	}

	if n != len(b) {
		return fmt.Errorf(
			"writing pre-run-action was incomplete - wrote %d bytes, expected %d", n, len(b))
	}

	klog.Infof("Persisted next pre-run action '%s' to %s", string(b), preRunActionFilepath)
	return nil
}
```

I have no problem with that, I'll make that change.

What bothers me (and that's a problem with current impl as well) is this doesn't test if the dir gets created and I already had couple surprises with unit tests vs real system :D 
For this, it would be useful to have fs abstraction like spf13/afero but I don't like how it puts more code between us and system...
Maybe overriding dest dirs is the way, some dir in `/tmp` for unit testing
Or, maybe we don't need as much focus on unit tests :see_no_evil:  :smiling_imp:

Reply:
Also: I don't see how can we unit test `exec.Command(cp --reflink)`, so I guess what you proposed and what's in code snippet I posted in comment above is good middle ground?
Use io.{Writer,Reader} where possible, but don't get crazy about coverage - instead of making it unreasonably testable by obscuring code, better invest in integration/e2e tests.
It doesn't apply to the pure functions, like "is data migration needed? Is data migration allowed?" - these should get all data as an input and produce output and given they're (should be) side effect, tested to the max

Reply:
We also have to be careful about adding another dependency (I haven't looked at what afero brings in and whether it is compatible with all of the kube stuff).

The nice aspect of defining an interface is we can make the semantics match its use (get, set, etc. instead of read, write, ...). We could also test the filesystem implementation independently of how it's used, although we still have a challenge with introducing mock write failures.

Is there a reliable way for the test cases to ensure they restore the original state of the function pointers they replace, so that after any given test is done we know the "real" implementation is restored?

Reply:
> Is there a reliable way for the test cases to ensure they restore the original state of the function pointers they replace, so that after any given test is done we know the "real" implementation is restored?

Two ideas

This one doesn't give any advantage over `persist(io.Writer)` while looking convoluted 
```go
func (nb *preRunAction) Persist() error {
	return nb.persist(func() (io.Writer, error) {
		if err := config.CreateDir(config.BaseDataDir); err != nil {
			return nil, err
		}
		return os.OpenFile(preRunActionFilepath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, filePerm)
	})
}

func (nb *preRunAction) persist(writerGetter func() (io.Writer, error)) error {
	w, err := writerGetter()
	if err != nil {
		return err
	}

	b, err := json.Marshal(nb)
	if err != nil {
		return err
	}
	n, err := w.Write(b)
	if err != nil {
		return err
	}

	if n != len(b) {
		return fmt.Errorf(
			"writing pre-run-action was incomplete - wrote %d bytes, expected %d", n, len(b))
	}

	klog.Infof("Persisted next pre-run action '%s' to %s", string(b), preRunActionFilepath)
	return nil
}
```

Second 
```go
var getFileWriter = getFileWriterImpl

func getFileWriterImpl() (io.Writer, error) {
	if err := config.CreateDir(config.BaseDataDir); err != nil {
		return nil, err
	}
	return os.OpenFile(preRunActionFilepath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, filePerm)
}
```
and following in test to restore original impl (it could also stored it within the test, but this felt prettier)
```go
defer func() {
	getFileWriter = getFileWriterImpl
}()
```

Reply:
If we can make it better, I have no problems with that - feel free to experiment

Reply:
Using defer seems pretty straightforward. Maybe we can define a utility function so all the tests can force reset all of the replaceable functions reliably.